src/candid.ts

```diff
@@ -0,0 +1,163 @@
+  const einBaseFieldId = await getEinBaseFieldId(baseUrl, token);
+  const proposals = await getProposals(baseUrl, token);
+
+  const eins = new Set(proposals.entries
+    .flatMap((proposal) => proposal.versions[0]?.fieldValues)
+    .filter(<T>(x: T | undefined): x is T => typeof x !== 'undefined')
+    .filter(({ applicationFormField }) => applicationFormField.baseFieldId === einBaseFieldId)
+    .map(({ value }) => value)
+    .filter(isValidEin));
+  return [...eins];
+};
+
+const candid: CommandModule = {
+  command: 'candid',
+  describe: 'Interact with the Candid Premier API',
+  builder: (y) => (y
+    .option('candid-api-key', {
+      describe: 'Candid Premier API key; get from account management at https://dashboard.candid.org/',
+      demandOption: true,
+      global: true,
+      type: 'string',
+    })
+    .command(
+      'lookup <ein>',
+      'Fetch and display information about an organization by its EIN',
+      {
+        'output-file': {
+          alias: 'write',
+          describe: 'Write organization information to the specified JSON file',
+          normalize: true,
+          type: 'string',
+        },
+      },
+      async (args) => {
+        const result = await getCandidProfile(
+          args.candidApiKey as string,
+          args.ein as string,
+        );
+
+        if (typeof args.outputFile === 'string') {
+          await writeFile(
+            args.outputFile,
+            JSON.stringify(result, null, 2),
+          );
+          logger.info(`Wrote Candid data for ${args.ein as string} to ${args.outputFile}`);
+        } else {
+          logger.info('Candid result', { result });
+        }
+      },
+    )
+    .command(
+      'update <ein>',
+      'Fetch information about an organization by its EIN, and upload to the PDC',
+      oidcOptions,
+      async (args) => {
+        const token = await getToken(
+          args.oidcBaseUrl as string,
+          args.oidcClientId as string,
+          args.oidcClientSecret as string,
+        );
+        const data = await getCandidProfile(args.candidApiKey as string, args.ein as string);
+        await postPlatformProviderData(
+          args.pdcApiBaseUrl as string,
+          token,
+          args.ein as string,
+          'Candid',
+          data,
+        );
+        logger.info(`Wrote data for ${args.ein as string} to PDC`);
+      },
+    )
+    .command(
+      'update-all',
+      'Update Candid profiles for all PDC proposals',
+      oidcOptions,
+      async (args) => {
+        const token = await getToken(
+          args.oidcBaseUrl as string,
+          args.oidcClientId as string,
+          args.oidcClientSecret as string,
+        );
+        const eins = await getEinsFromPdc(args.pdcApiBaseUrl as string, token);
+        logger.info(`Found ${eins.length} valid EINs to look up in Candid`);
+
+        for (let i = 0; i < eins.length; i += 1) {
+          /* eslint-disable no-await-in-loop -- rate limit ourselves so Candid doesn't */
```

Comment:
If you have a better option than an old-school `for` loop that won't fire off every request at the same time, please share!

Reply:
I do!  I bet you're not gonna like it though.

The only way I've found to have a series of promises that wait on one another is to use an async reduce.  Here's an example: https://github.com/PermanentOrg/sftp-service/blob/a1a0d740c7100e6689ead2953ddc41228db9f142/src/classes/PermanentFileSystem.ts#L346-L359

I believe that a for with an await can, in some cases, result in blocking code / starve the event loop, so I would recommend avoiding it (though since you're awaiting on I/O here I think it ends up being OK).

Separately from all that though, what do you think of updating `getCandidProfile` to `getCandidProfiles` and having it take an array / be responsible for rate limiting and negotiating with chunking against the API.

For an example of why I think that breakdown would be a cleaner separation of responsibilities: in Charity Navigator the `get` endpoint they provide allows you to send multiple EINs so I will be having `getCharityNavigatorProfiles` that encapsulates the negotiation with the API appropriately.

Reply:
> The only way I've found to have a series of promises that wait on one another is to use an async reduce.

Hah, I saw that while I was exploring my options. Do you think that's *better*? I ask sincerely, because I don't know if I have an opinion one way or another - as we've discussed, I tend to find `reduce`rs hard to read / reason about, but having to silence eslint warnings is also Not Great. 

> I believe that a for with an await can, in some cases, result in blocking code / starve the event loop, so I would recommend avoiding it (though since you're awaiting on I/O here I think it ends up being OK).

Can you say more? I'm not sure I follow!

> Separately from all that though, what do you think of updating `getCandidProfile` to `getCandidProfiles` and having it take an array / be responsible for rate limiting and negotiating with chunking against the API.

Sure, that sounds like a good idea to me!

Reply:
> Do you think that's better?

I do prefer the use of reduce to a for loop, but I appreciate that reduce is more complex.  What I *really* wish is that something like Promise.sequence() existed and then you could just map the elements to a list of promises.

You probably saw this but holy moly (I think I prefer reduce to this, but hey, it's an option):

```
function runSerial(tasks) {
  var result = Promise.resolve();
  tasks.forEach(task => {
    result = result.then(() => task());
  });
  return result;
}
```
from https://stackoverflow.com/a/31070150/159522

> Can you say more? I'm not sure I follow!

If those links (specifically the first one / https://snyk.io/blog/nodejs-how-even-quick-async-functions-can-block-the-event-loop-starve-io/) don't explain it well enough then I'd probably be better of explaining by voice.  It involves some pretty low level base knowledge of the node event loop (which you may have! but I'd rather navigate that by voice).

Basically using await in a for loop does NOT mean that your loop is necessarily non-blocking; you can have a situation where microtasks generate more microtasks (and the way the event loop works, the loop doesn't continue until that queue is completely depleted).  I don't believe that applies here because of the use of I/O in the await, but by disabling the eslint rule we open the doors for unwittingly allowing it.

Reply:
> Basically using await in a for loop does NOT mean that your loop is necessarily non-blocking; you can have a situation where microtasks generate more microtasks (and the way the event loop works, the loop doesn't continue until that queue is completely depleted). I don't believe that applies here because of the use of I/O in the await, but by disabling the eslint rule we open the doors for unwittingly allowing it.

Ah, that is informative, thank you!

To expand on what you said, not only are we doing I/O rather than heavy computations, I think we also *don't care about the event loop* in this context - right? This is a single-purpose batch script being run until completion, so I don't think there are any events we would care about - with the possible exception (ha) of system signals, like `SIGINT` (often generated by `Ctrl-C`), which I don't think go through the event loop, but I am not sure about that.

All of that said, this situation is specifically called out in the [documentation for the `no-await-in-loop` rule](https://eslint.org/docs/latest/rules/no-await-in-loop):

> Or, loops may be used to prevent your code from sending an excessive amount of requests in parallel. In such cases it makes sense to use `await` within a loop and it is recommended to disable the rule via a standard ESLint disable comment.

So maybe I should get over my own distaste of `eslint-disable` comments?

>  What I really wish is that something like `Promise.sequence()` existed and then you could just map the elements to a list of promises.

Oh, that's a good call - is there a library we could lean on for that? I see a few options on npm, but none of them are very recent; I haven't dug in yet to see if that's because they're mature or abandoned.

Reply:
> don't care about the event loop in this context

Iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii get what you're saying in theory but it's hard for me to imagine feeling OK with code that locks out I/O (we *do* care about the event loop since, for instance, network events are events too and we need to be able to consume network responses!) But yeah it's a moot point since we're not doing it.  Blocking the event loop means node isn't able to revisit async code blocks, for instance.  Idk!

Basically "don't block the event loop" is one of very few top level guides published by node: https://nodejs.org/en/docs/guides/dont-block-the-event-loop

> this situation is specifically called out

Hah the explicit callout is pretty strong evidence that it's a totally fine way to go.  (bottom line is that in a vaccum I'd use reduce, hence the example I linked to, but I think the foreach is fine)

I +1 the distaste to lint rule avoidance, but being able to say in a comment "this is explicitly suggested by the rule itself" is as iron clad as one can get.

Reply:
( oh and yeah I couldn't find any libraries for it last time I looked :( -- I think folks were like "just... use reduce!")

Reply:
So @slifty did we land on "this is fine, actually"? Or would you still prefer to use reduce?

Reply:
I think this is fine!